text/0013-abort-pending-updates.md

```diff
@@ -0,0 +1,143 @@
+- Title: Abort pending updates
+- Start Date:
+- Specification PR: https://github.com/meilisearch/specifications/pull/13
+- MeiliSearch Issue: 
+    - https://github.com/meilisearch/MeiliSearch/issues/1104
+- Parent Spec: https://github.com/meilisearch/specifications/pull/12
+
+
+# Abort pending updates
+
+## 1. Feature Description and Interaction
+
+### Summary
+
+Allow to abort messages in the update queue or to empty the entire queue. An aborted update will have the status `aborted`.
+
+Questions not answered:
+- We have an attribute on update info `processedAt`. What becomes this attribute if we add the possibility to abort an update? We already use this term in the case of a failed update.
+
+### Motivation
+
+The update queue is convenient because it allows us to serialize updates and assure the [ACID property](https://en.wikipedia.org/wiki/ACID) of MeiliSearch. It also causes some problems. As the queue message is an append-only queue, it is impossible to remove an update from this queue. The updates can be large and can take a long time to index. Sometimes, it would be convenient to abort a message in this queue or empty it. 
+
+### Additional Materials
+
+As explained in the [parent spec](https://github.com/meilisearch/specifications/pull/12), It's quite rare to have this kind of update queue. Elasticsearch or Algolia, who have this kind of update queue, don't give the abort or clear their 'tasks' queues. 
+
+### Explanation
+
+Two new routes and a new status (`aborted`) must be added to make this feature possible. 
+
+An aborted update will stay at its place in the queue. So the typical answer of the get all update status route will be: 
+
+```json
+[
+    {
+        "status": "processed",
+        "updateId": 1,
+        ...
+    },
+    {
+        "status": "failed",
+        "updateId": 2,
+        ...
+    },
+    {
+        "status": "processed",
+        "updateId": 3,
+        ...
+    },
+    {
+        "status": "pending",
+        "updateId": 4,
+        ...
+    },
+    {
+        "status": "aborted",
+        "updateId": 5,
+        ...
+    },
+    {
+        "status": "pending",
+        "updateId": 6,
+        ...
+    },
+    ...
+]
+```
+
+the update status document will look like:
+
+```json
+{
+  "status": "aborted",
+  "updateId": 12,
+  "type": {
+    "name": "DocumentsAddition",
+    "number": 4
+  },
+  "duration": 0,
+  "enqueuedAt": "2019-12-07T21:16:09.623944Z",
+  "processedAt": "2019-12-07T21:16:09.703509Z" // date of abortion
+}
+```
+
+#### HTTP API
```

Comment:
There is one thing that I like a little less but it could be only me. 

## Resftullness
I don't like the fact that the `delete` route is not RESTful in the sense that the delete method works as follow: 

>As the name applies, DELETE APIs are used to delete resources (identified by the Request-URI).

>DELETE operations are idempotent. If you DELETE a resource, **it’s removed from the collection of resources**. Repeatedly calling DELETE API on that resource will not change the outcome – however, calling DELETE on a resource a second time will return a 404 (NOT FOUND) since it was already removed. Some may argue that it makes the DELETE method non-idempotent. It’s a matter of discussion and personal opinion.

For us, this is not true. We only delete a resource if it a specific condition in met. If it is not, nothing is deleted and an error is returned.

## Comparison with ES async

I have looked at what elasticSearch does in their [async-search](https://www.elastic.co/guide/en/elasticsearch/reference/current/async-search.html) route. 

This is there delete route: 
> Delete async searched
You can use the delete async search API to manually delete an async search by ID. **If the search is still running, the search request will be cancelled. Otherwise, the saved search results are deleted**.

They take the party to say: oke you are using the delete route which means you want to delete this resource. 
- If it is `running`, we abord it 
- If it is `done`, we delete it

In our case it would translate the following way: 
- if it is `enqueued`, we abord/remove it
- if it is `running`, we abord it (for later I have red but it would be that)
- if it is `done`, we delete it.

But the reality of what you suggest is 
- If it is `enqueued`, we abord/remove it. 
- if it is `running`: error
- if it is `done`: error

I understand that those errors are made to make the user aware of how `DELETE /update` works, but as an end result it is less intuitive because more developers will expect it to work as a regular RESTfull delete.

## Suggestion

I would change the name of the route
- Route: `DELETE /indexes/:index_uid/updates/enqueued/:update_id`
- Route: `DELETE /indexes/:index_uid/updates/enqueued`

In this case we become restful and intuitive. The errors would also be way more understandable in this context.

## Conclusion 

So my general concern is that this route is not RESTfull and also quiet unclear as intuitively you would think that using DELETE on the update route would remove an update from the update array which is not the case.

And the only example I found (elasticsearch) is restful. 

Of course, I do not want to postpone any development. So these could be arguments for a later rework.

Reply:
Agreed!

Either we are ABORTING it (which is a status update and might be a PUT request if RESTFUL) or we are DELETING it (in which case it is a DELETE request if restful) and it won't be possible to find it afterwards!

Reply:
So in this case you find it more clear if we propose a PUT on `/indexes/:index_uid/updates/pending/:update_id/abort`?

Reply:
I agree with @eskombro, we should consider this, but without postponing the development as @bidoubiwa mentioned 👍 
I don't really like the huge route name, except if it's the only solution, but we can also think about routes like this (it might be what @eskombro suggested):
- `PUT /indexes/:index_uid/updates`
- `PUT /indexes/:index_uid/updates/:update_id`

Because here we update the `update` resource with a new status

Reply:
Btw being REST is not an obligation for everything
- our routes are not all REST: the `POST` route to update all the settings. And also we use the `PUT` route to update documents but, I read different articles that says we should use `PATCH` instead of `PUT`. I also read the contrary!
- the stripe API is not rest at all -> they only use `POST`, `GET` and `DELETE`, but they are really easy to use :)

What I want to say is not "we should not take this into account" (as I said above, I even suggest something) but we should not abandon "easy to use" for "being RESTful".

Reply:
> I agree with @eskombro, we should consider this, but without postponing the development as @bidoubiwa mentioned 👍
> I don't really like the huge route name, except if it's the only solution, but we can also think about routes like this (it might be what @eskombro suggested):
> 
>     * `PUT /indexes/:index_uid/updates`
> 
>     * `PUT /indexes/:index_uid/updates/:update_id`
> 
> 
> Because here we update the `update` resource with a new status

This is exactly what I had in mind. And yes, RESTFUL is not an obligation, just worth considering :)

Reply:
To be clear, for you, it's logical and understandable that if you do `PUT /indexes/:index_uid/updates/:update_id`, that will abort your update?

Reply:
>To be clear, for you, it's logical and understandable that if you do PUT /indexes/:index_uid/updates/:update_id, that will abort your update?

No it is not, I think we should add a `/abort` to because **PUT does not imply abortion**
. 
I agree with the fact that `DELETE` implies a deletion, which is also not what we are doing, so `delete` is not a good idea.

 
So I would go with @qdequele suggestion (if it was a suggestion)

`PUT /indexes/:index_uid/updates/pending/:update_id/abort`

or, if we want to prepare for a future where tasks that are being `processed` we should go for this:

`PUT /indexes/:index_uid/updates/:update_id/abort`

This will do the following:

**v1**
- If it is **enqueued**: we change status to aborted.
- if it is **processing**: `Error: Cannot abord task that is being processed`
- if it is done: `Error: Cannot abord task that has already been processed`

**v2**
- If it is **enqueued**, We change status to aborted.
- if it is **processing**: We abord and change the status to aborted
- if it is done: `Error: Cannot abord task that has already been processed`

Reply:
🧐. BTW I like the fact that if you DELETE an processed update/task, this one will be really deleted. What you proposed @bidoubiwa previously, delete a pending update, change it to aborted, delete a processed update really delete it.

Reply:
1. Yes, using the PUT option would imply passing a parameter that explicitely asks for abortion. Yet, as you point out, this is not clear in the route.

2. The DELETE option is better. But if you delete an "update" that has been processed, this would add too much complications because the history wouldn't be consistent with the data.

As pointed out before, we don't need to be 100% restful and can create an /abort route.

Reply:
> The DELETE option is better. But if you delete an "update" that has been processed, this would add too much complications because the history wouldn't be consistent with the data.

MeiliSearch does not guarantee the consistency between the data and the update history otherwise the update history would not be empty when you import a dump. I mentioned this inconsistency during dump implementation and this is not something that MeiliSearch guarantees. The updates are for consultations: MeiliSearch is not introduced as a real DB and the update history cannot be compared to migration states/files. This is what we said after the meeting about the dumps.

Reply:
After an internal discussion we choose to:

- Renaming updates to tasks
The task is more generic, will allow us to put inside the Dumps, etc..

- Renaming the status enqueued to pending
But we keep the enqueuedAt

- Renaming abort to cancel
Abort means already in process/cancel means not yet started

- Use PUT to change the status of a task
Based on the Stripe API. (edited) 

- Keep always all fields in the object